Detectors/CPV/calib/CPVCalibWorkflow/src/CPVPedestalCalibDevice.cxx

```diff
@@ -0,0 +1,253 @@
+}
+
+void CPVPedestalCalibDevice::sendOutput(DataAllocator& output)
+{
+  // extract CCDB infos and calibration objects, convert it to TMemFile and send them to the output
+  // TODO in principle, this routine is generic, can be moved to Utils.h
+  // using clbUtils = o2::calibration::Utils;
+  if (mUpdateCCDB || mForceUpdate) {
+    // prepare all info to be sent to CCDB
+    o2::ccdb::CcdbObjectInfo info;
+    auto image = o2::ccdb::CcdbApi::createObjectImage(mPedestals.get(), &info);
+
+    auto flName = o2::ccdb::CcdbApi::generateFileName("Pedestals");
+    info.setPath("CPV/Calib/Pedestals");
+    info.setObjectType("Pedestals");
+    info.setFileName(flName);
+    // TODO: should be changed to time of the run
+    const auto now = std::chrono::system_clock::now();
+    long timeStart = std::chrono::duration_cast<std::chrono::seconds>(now.time_since_epoch()).count();
+    info.setStartValidityTimestamp(timeStart);
+    info.setEndValidityTimestamp(99999999999999);
+    std::map<std::string, std::string> md;
+    info.setMetaData(md);
+
+    LOG(INFO) << "Sending object CPV/Calib/Pedestals";
+
+    header::DataHeader::SubSpecificationType subSpec{(header::DataHeader::SubSpecificationType)0};
+    output.snapshot(Output{o2::calibration::Utils::gDataOriginCLB, o2::calibration::Utils::gDataDescriptionCLBPayload, subSpec}, *image.get());
+    output.snapshot(Output{o2::calibration::Utils::gDataOriginCLB, o2::calibration::Utils::gDataDescriptionCLBInfo, subSpec}, info);
+  }
+  //Anyway send change to QC
+  LOG(INFO) << "[CPVPedestalCalibDevice - run] Writing ";
+  output.snapshot(o2::framework::Output{"CPV", "PEDDIFF", 0, o2::framework::Lifetime::Timeframe}, mPedDiff);
+
+  //Write pedestal distributions to calculate bad map
+  std::string filename = mPath + "CPVPedestals.root";
+  TFile f(filename.data(), "RECREATE");
+  mMean->Write();
+  f.Close();
+}
+
+void CPVPedestalCalibDevice::calculatePedestals()
+{
+
+  mPedestals.reset(new Pedestals());
+
+  //Calculate mean of pedestal distributions
+  for (unsigned short i = mMean->GetNbinsX(); i > 0; i--) {
+    TH1D* pr = mMean->ProjectionY(Form("proj%d", i), i, i);
+    short pedMean = std::min(255, int(pr->GetMean()));
+    pr->Delete();
+    mPedestals->setPedestal(i - 1, pedMean);
+  }
+}
+
+void CPVPedestalCalibDevice::checkPedestals()
+{
+  //Compare pedestals to current ones stored in CCDB
+  //and send difference to QC to check
+  if (!mUseCCDB) {
+    mUpdateCCDB = true;
+    return;
+  }
+  // //TODO:
+  // //Get current map
+  // int nChanged=0;
+  // for(short i=o2::cpv::Geometry::kNCHANNELS; --i;){
+  //   short dp=mPedestals.getPedestal(i)-oldPed.getPedestal(i);
+  //   mPedDiff[i]=dp ;
+  //   if(abs(dp)>1){ //not a fluctuation
+  //     nChanged++;
+  //   }
+  // }
+  // if(nChanged>kMinorChange){ //serious change, do not update CCDB automatically, use "force" option to overwrite
+  //   mUpdateCCDB=false;
+  // }
+  // else{
+  //   mUpdateCCDB=true;
+  // }
+}
```

Comment:
Are those commented methods to be refined later? If not, better to suppress the dead code.

Reply:
Yes, I will implement them together with sending/reading to CCDB.